go/tools/builders/compile.go

```diff
@@ -0,0 +1,90 @@
+// Copyright 2017 The Bazel Authors. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// compile compiles .go files with "go tool compile". It is invoked by the
+// Go rules as an action.
+package main
+
+import (
+	"fmt"
+	"go/build"
+	"log"
+	"os"
+	"os/exec"
+	"path/filepath"
+)
+
+func run(args []string) error {
+	// process the args
+	if len(args) < 2 {
+		return fmt.Errorf("Usage: compile gotool [sources] -- <extra options>")
+	}
+	gotool := args[0]
+	args = args[1:]
+	sources := []string{}
+	goargs := []string{}
+	bctx := build.Default
+	bctx.CgoEnabled = true
+	for i, s := range args {
+		if s == "--" {
+			if i < len(args) {
+				goargs = args[i+1:]
+			}
+			break
+		}
+		sources = append(sources, s)
+	}
+	// filter the soure list to the ones that apply
```

Comment:
Typo in "soure". But consider rewording to "Apply build constraints to the source list".